posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md

```diff
@@ -0,0 +1,209 @@
+
+- SciPy currently lacks any developer documentation for CI providers. Documenting the CI systems and the build matrix is necessary for developers to cross-reference them in future improvements. We can also document how to emulate CI runs locally through specific tools.
+
+## Implementing the local CI run
+
+The project's first step was to ensure that we could run the GitHub Actions CI locally to ensure faster iterations on the failures we encountered. GitHub does not feature an official tool to make this possible; hence we had to look at third-party alternatives like Act.
+
+`act` is a tool offered by Nektos which provides a handy way to run GitHub Actions locally. It gives a quick way to validate your changes on the CI locally, without committing/pushing your changes to the workflows to trigger and validate them. It leads to fast feedback and compatibility as a local task runner to validate all our CI jobs.
+
+You can try out `act` yourself by cloning [Ralf’s fork](https://github.com/rgommers/scipy) and referencing the official documentation for the [`act` usage](https://github.com/rgommers/scipy/blob/meson/ACT-USAGE.md). It keeps the GitHub Action notifications clutter-free and reduces the chance of accidentally triggering any failure during debugging.
+
+Here is `act` in action with the `meson` branch over [Ralf’s fork](https://github.com/rgommers/scipy):
+
+<p align="center">
+    <img
+     alt="A snapshot from the Visual Studio Code demonstrating the usage of `act` locally to test out the GitHub Action workflows. In the picture, we are running the `test_meson` workflow which runs the CI job locally to build, install and test SciPy locally."
+     src="/images/2021/10/re-engineering-cicd-pipelines-for-scipy/scipy-github-actions-act-usage.png">
+    <i>Usage of `act` with GitHub Actions for SciPy</i>
+</p>
+
+To ensure that we can test the GitHub Actions locally, we had to make a lot of tweaks to ensure that `act` plays nicely with local workflows. [PR #57](https://github.com/rgommers/scipy/pull/57) containing the commits for the tweaks and the overall developer documentation for the `act` usage.
+
+## Generating dynamic files in the build
+
+During the Meson build process, we generate some of the dynamic files during the process. The overall plan was to generate these files programmatically and incorporate this into the build in the most sensible way. The files that we were looking to generate dynamically were:
+
+- `__config__.py`: This file is generated by NumPy and should use a version of NumPy at build time that is compatible with all potential runtime versions of NumPy.
+
+- `version.py`: This file is dependent on the Git tags. We were looking to instantiate a file in the build directory that represents the current one.
```

Comment:
"the Git tags" -> "Git commit hash"

Reply:
"We were looking to instantiate a file in the build directory that represents the current one." ->

"We needed to generate a file in the build directory instead of in-place, because Meson does not allow writing output into the source tree that is being built."